stdlib/experimental/prometheus/scrape.go

```diff
@@ -0,0 +1,334 @@
+	spec, ok := prSpec.(*ScrapePrometheusProcedureSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", prSpec)
+	}
+
+	PrometheusIterator := PrometheusIterator{id: dsid, spec: spec, administration: a}
+
+	return execute.CreateSourceFromDecoder(&PrometheusIterator, dsid, a)
+}
+
+type PrometheusIterator struct {
+	id             execute.DatasetID
+	administration execute.Administration
+	spec           *ScrapePrometheusProcedureSpec
+	reader         *execute.RowReader
+
+	metrics []Metric
+	met     Metric
+	i       int
+	url     string
+	resp    *http.Response
+}
+
+// Metric stores the fields that we need to construct Table
+type Metric struct {
+	Name      string                 `json:"name"`
+	Tags      map[string]string      `json:"tags"`
+	Fields    map[string]interface{} `json:"fields"`
+	Timestamp time.Time              `json:"timestamp"`
+}
+
+func (p *PrometheusIterator) Connect(ctx context.Context) error {
+	p.url = p.spec.URL // for easier reference
+
+	resp, err := http.Get(p.url)
+	p.resp = resp
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Parse the entire body into a slice of Metrics
+	err = p.parse(resp.Body, resp.Header)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *PrometheusIterator) parse(r io.Reader, header http.Header) (err error) {
+	var parser expfmt.TextParser
+	now := time.Now()
+
+	mediatype, params, err := mime.ParseMediaType(header.Get("Content-Type"))
+	if err != nil {
+		return err
+	}
+
+	// Prepare output
+	metricFamilies := make(map[string]*dto.MetricFamily)
+	if mediatype == "application/vnd.google.protobuf" &&
+		params["encoding"] == "delimited" &&
+		params["proto"] == "io.prometheus.client.MetricFamily" {
+		for {
+			mf := &dto.MetricFamily{}
+			if _, err := pbutil.ReadDelimited(r, mf); err != nil {
+				if err == io.EOF {
+					break
+				}
+				return fmt.Errorf("reading metric family protocol buffer failed: %s", err)
+			}
+			metricFamilies[mf.GetName()] = mf
+		}
+	} else {
+		metricFamilies, err = parser.TextToMetricFamilies(r)
+		if err != nil {
+			return fmt.Errorf("reading text format failed: %s", err)
+		}
+	}
+
+	p.metrics = make([]Metric, 0)
+
+	// read metrics
+	for name, family := range metricFamilies {
+		for _, m := range family.Metric {
+			// read tags
+			tags := makeLabels(m)
+
+			// read fields
+			fields := make(map[string]interface{})
+			fields = getNameAndValue(m)
+
+			// Store data needed to build flux Table
+			if len(fields) > 0 {
+				tm := now
+				if m.TimestampMs != nil && *m.TimestampMs > 0 {
+					tm = time.Unix(0, *m.TimestampMs*1000000)
+				}
+				met := Metric{
+					Timestamp: tm,
+					Tags:      tags,
+					Fields:    fields,
+					Name:      name,
+				}
+				p.metrics = append(p.metrics, met)
+			}
+		}
+	}
+	return nil
+}
+
+func (p *PrometheusIterator) Fetch(ctx context.Context) (bool, error) {
+	// Iterate over all Metrics in the slice
+	p.i++
```

Comment:
Is there an off-by-one issue here?  It seems like the metric in `p.metrics[0]` will never be returned.

Reply:
Also I think Adam alluded to this in one of his earlier comments, but you could just store `p.i` and `p.metrics` as state, and then index into the slice via `p.metrics[p.i]` in the `Decode` method.  Then you would avoid having to copy a `Metrics` object every time you decode.

Reply:
Thanks for pointing out the off by one! And yeah, thanks for helping me understand the copying issue. Fixing.